programs/bpf_loader/src/syscalls.rs

```diff
@@ -2299,134 +2499,1722 @@ fn call<'a>(
     Ok(SUCCESS)
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
-    use solana_rbpf::{
-        ebpf::HOST_ALIGN, memory_region::MemoryRegion, user_error::UserError, vm::Config,
-    };
-    use solana_sdk::{
-        bpf_loader,
-        fee_calculator::FeeCalculator,
-        hash::hashv,
-        process_instruction::{MockComputeMeter, MockInvokeContext, MockLogger},
-    };
-    use std::str::FromStr;
-
-    const DEFAULT_CONFIG: Config = Config {
-        max_call_depth: 20,
-        stack_frame_size: 4_096,
-        enable_instruction_meter: true,
-        enable_instruction_tracing: false,
+/// Bignum data cost calculator
+macro_rules! calc_bignum_cost {
+    ($input_self:expr, $input_size:expr) => {
+        $input_self.cost
+            + ($input_self.word_cost as f64
+                * ($input_size as f64 / $input_self.word_div_cost as f64).ceil()) as u64
     };
+}
 
-    macro_rules! assert_access_violation {
-        ($result:expr, $va:expr, $len:expr) => {
-            match $result {
-                Err(EbpfError::AccessViolation(_, _, va, len, _)) if $va == va && len == len => (),
-                _ => panic!(),
-            }
-        };
+/// Transposes a bn::BigNum into an FfiBignumber array of bytes
+fn bignum_to_ffibignumber(
+    bignum: &BigNum,
+    bnffi: &mut FfiBigNumber,
+    loader_id: &Pubkey,
+    memory_mapping: &MemoryMapping,
+    result: &mut Result<u64, EbpfError<BpfError>>,
+) {
+    // Get the BigNum vector and size
+    let big_number_bytes = bignum.to_vec();
+    let big_number_len = big_number_bytes.len() as u64;
+    (*bnffi).is_negative = bignum.is_negative();
+    // Fail if we can't support the full array of bytes
+    if big_number_len as u64 > bnffi.data_len {
+        *result = Err(SyscallError::BigNumberToBytesError.into())
+    } else {
+        // Get the output array pointer
+        let bn_buffer_out = question_mark!(
+            translate_slice_mut::<u8>(memory_mapping, bnffi.data, bnffi.data_len, loader_id, true),
+            result
+        );
+        // Zero result
+        if big_number_len == 0 {
+            bn_buffer_out[0] = 0u8;
+            (*bnffi).data_len = 1;
+        } else if big_number_len == bnffi.data_len {
+            (*bn_buffer_out).copy_from_slice(&big_number_bytes);
+        } else {
+            let mut index = 0;
+            big_number_bytes.iter().for_each(|byte| {
+                bn_buffer_out[index] = *byte;
+                index += 1;
+            });
```

Comment:
Why can't this just be `(*bn_buffer_out).copy_from_slice(&big_number_bytes);`? You're open-coding a byte-copy here.

Reply:
@seanyoung - We do not know for certain what the resulting size of the BigNum will be in bytes. We use an approximation. `copy_from_slice` will fail unless the two vectors are the same size.

Reply:
You could do: 
```rust
bn_buffer_out[..big_number_bytes.len()].copy_from_slice(&big_number_bytes)
```
Also do we need to bounds check to ensure that the user can't pass slices with lengths that would cause an exception?

Reply:
> > > I have a concern that this might be used as a denial of service attack. As far as I can make out, openssl supports potentially huge numbers: INT_MAX / 4 bits, from reading the openssl code -- I have not tested this.
> > > If someone creates two huge bignums and then calls sol_bignum_div() on them, will the solana process spends hours working out the answer?
> > > I have not tried this out.
> > 
> > 
> > This should be blocked by the ComputeBudget. Know if the division is linear time over length? We will probably have to tune the compute cost numbers
> 
> I don't know. Should we put limits in just as belt and braces? E.g. bignums can't be more than 8192 bit.

@seanyoung - To clarify: Are you saying that the result of any operation can not be bigger than 8192? We are good to make that change and change it if it's a problem.

Reply:
@jackcmay 
```
bn_buffer_out[..big_number_bytes.len()].copy_from_slice(&big_number_bytes)
```
I've tried that in the past when you first suggested, but...
```
the type `[u8]` cannot be indexed by `std::ops::RangeTo<u64>`
the trait `std::slice::SliceIndex<[u8]>` is not implemented for `std::ops::RangeTo<u64>`
required because of the requirements on the impl of `std::ops::Index<std::ops::RangeTo<u64>>` for `[u8]`rustcE0277
```

Reply:
@seanyoung @arthurgreef @jackcmay 
As per Jack, we will do the result checks in the program space before calling the syscall...

Reply:
This works:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=502deeefbc8cf016b6fa189f154f3a58

Need to make sure to check size, would be nice to put the size check in the line one above the slice copy to make thing extra clear

Reply:
There is a bound check that if the result of the function produces something larger than the result BigNumber buffer will fail...

Reply:
> > I don't know. Should we put limits in just as belt and braces? E.g. bignums can't be more than 8192 bit.
> 
> @seanyoung - To clarify: Are you saying that the result of any operation can not be bigger than 8192? We are good to make that change and change it if it's a problem.

Well I would say that both the inputs and the result  can't be larger than 8192, if @jackcmay agrees.

Reply:
Pushed change to above suggestion

Reply:
So, in some cases, there are two gates:
1. If the output buffer for the result is detected to exceed 1024 bytes (8192 b) then the syscall errors, which in some cases will prevent lots of math function cycles (e.g. mod_exp, exp, etc.) otherwise...
2. If the result of the operation would exceed the 1024 byte mark

So, no BigNumber should exist that is greater than 1024 bytes